src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,904 @@
-  renaming (suc to 1+_; _+_ to _+ℕ_; _*_ to _*ℕ_; _≤_ to _≤ℕ_)
-import Data.Nat.Properties as ℕₚ
-open import Data.Product using (proj₁; proj₂)
-open import Function using (_∘_)
-open import Relation.Binary
-open import Relation.Binary.Consequences
-open import Relation.Binary.PropositionalEquality
-  using (_≡_; _≢_; refl; sym; isEquivalence; resp₂; decSetoid)
-open import Relation.Nullary using (yes; no)
+open import Algebra using (Magma; Semigroup; Monoid; CommutativeMonoid;
+                                             Semiring; CommutativeSemiring)
+import      Algebra.FunctionProperties as FuncProp
+open import Algebra.Structures using
+            (IsMagma; IsSemigroup; IsMonoid; IsCommutativeMonoid; IsSemiring;
+             IsSemiringWithoutAnnihilatingZero; IsCommutativeSemiring
+            )
+open import Data.Bin.Base using (Bin; _≟_; suc; pred; _+_; _*_; 2*; 2^; size;
+                                                                1#; 2#; toℕ)
+open import Data.Nat using (ℕ; s≤s)
+                     renaming (suc to 1+_; pred to predₙ; _+_ to _+ₙ_;
+                               _*_ to _*ₙ_; _∸_ to _∸ₙ_; _≤_ to _≤ₙ_)
+            -- to distinguish these operators from their Bin versions
 
-------------------------------------------------------------------------
--- (Bin, _≡_) is a decidable setoid
-
-1#-injective : ∀ {as bs} → as 1# ≡ bs 1# → as ≡ bs
-1#-injective refl = refl
-
-infix 4 _≟_ _≟ₑ_
-
-_≟ₑ_ : ∀ {base} → Decidable (_≡_ {A = Expansion base})
-_≟ₑ_ []       []       = yes refl
-_≟ₑ_ []       (_ ∷ _)  = no λ()
-_≟ₑ_ (_ ∷ _) []        = no λ()
-_≟ₑ_ (x ∷ xs) (y ∷ ys) with x Fin.≟ y | xs ≟ₑ ys
-... | _        | no xs≢ys = no (xs≢ys ∘ proj₂ ∘ ∷-injective)
-... | no  x≢y  | _        = no (x≢y   ∘ proj₁ ∘ ∷-injective)
-... | yes refl | yes refl = yes refl
-
-_≟_ : Decidable {A = Bin} _≡_
-0#    ≟ 0#    = yes refl
-0#    ≟ bs 1# = no λ()
-as 1# ≟ 0#    = no λ()
-as 1# ≟ bs 1# with as ≟ₑ bs
-... | yes refl  = yes refl
-... | no  as≢bs = no (as≢bs ∘ 1#-injective)
-
-≡-isDecEquivalence : IsDecEquivalence _≡_
-≡-isDecEquivalence = record
-  { isEquivalence = isEquivalence
-  ; _≟_           = _≟_
-  }
-
-≡-decSetoid : DecSetoid _ _
-≡-decSetoid = decSetoid _≟_
+open import Data.Nat.Properties as ℕp using (2m≢1+2n; module ≤-Reasoning)
+open import Data.Product    using (_,_; proj₁; ∃)
+open import Data.Sum        using (_⊎_; inj₁; inj₂)
+open import Function        using (_∘_; _$_; id; const; case_of_)
+open import Level           using () renaming (zero to 0ℓ)
+open import Relation.Binary using (Setoid; DecSetoid; IsDecEquivalence)
+open import Relation.Binary.Core using (_←→_)
+open import Relation.Binary.PropositionalEquality as P using
+                 (_≡_; _≢_; _≗_; refl; sym; trans; cong; cong₂; isEquivalence)
+open import Relation.Nullary          using (yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+
+open P.≡-Reasoning
+open ≤-Reasoning using () renaming (begin_ to begin≤_; _∎ to _end≤;
```

Comment:
Please don't rename the reasoning combinators.

Reply:
Fixed
-- by replacing usage of  ``≤-Reasoning``  with  ``subst (_≤ₙ cnt) eq``.
Because ``≤-Reasoning`` was used only a couple of times and with a small chain.